content/en/blog/2023-06-09-tales-of-open-source-vfscore.md

```diff
@@ -0,0 +1,414 @@
+```
+
+We see that a larger part of the source components that were part of Unikraft's `vfscore`, via OSv's `vfs`, are part of Prex's `vfs`: `vfs_mount.c`, `vfs_lookup.c`, `vfs_vnode.c`, `vfs_syscalls.c`.
+
+Prex features a [BSD style license](https://prex.sourceforge.net/doc/license.html), making its code reusable in OSv and then in Unikraft.
+
+Browsing the source code reveals that it was written from scratch.
+This is as far as we go with respect to source code origin.
+The only additional reference is mentioned in `vfs_bio.c`: the book [The Design of the UNIX Operating System (by Maurice Bach)](https://www.amazon.com/Design-UNIX-Operating-System/dp/0132017997).
+
+What we can determine is when in time did `vfs` get implemented in Prex, by checking [different versions](https://prex.sourceforge.net/downloads.html).
+As the description for version 0.5.0 says `File system`, we check that version and the previous one (0.4.3):
+
+```console
+$ wget http://prdownloads.sourceforge.net/prex/prex-0.5.0.tar.gz
+
+$ tar xf prex-0.5.0.tar.gz
+
+$ ls -F prex-0.5.0/user/
+arch/  bin/  include/  lib/  Makefile  sample/  server/  test/
+
+$ ls prex-0.5.0/user/server/fs/vfs/
+bio.c  conf.c  lookup.c  main.c  Makefile  mount.c  syscalls.c  tcb.c  vfs.h  vnode.c
+
+$ ls -F prex-0.5.0/user/server/fs/
+arfs/  devfs/  Makefile  ramfs/  vfs/
+
+$ wget http://prdownloads.sourceforge.net/prex/prex-0.4.3.tar.gz
+
+$ tar xf prex-0.4.3.tar.gz
+
+$ ls prex-0.4.3/user/
+arch/  bin/  include/  lib/  Makefile  sample/  test/
+```
+
+We that, indeed, the `vfs` implementation appeared in Prex 0.5.0, released in June 2007.
+Header files with `vfs`-related structure are present in Prex 0.4.3, but no implementation was yet present.
+
+## Full Story
+
+So we now have the full story of the origin of Unikraft's `vfscore`, based on the amazing process of reusing open source code.
+Initially implemented as part of the filesystem server in the [Prex real-time microkernel](https://prex.sourceforge.net/), it was then imported as [OSv's `vfs`](https://github.com/cloudius-systems/osv/tree/master/fs/vfs) and then added to [Unikraft's `vfscore`](https://github.com/unikraft/unikraft/tree/staging/lib/vfscore).
+
+The timeline is:
+
+- March 2005: Prex 0.1 is released by Kohsuke Ohtani
+- June 2007: Initial `vfs` implementation is added to Prex 0.5.0
+- October 2009: Latest version of Prex (0.9.0) is released
+- January 2013: Prex 0.9.0 `vfs` is imported in `OSv` as commit [`47ec18e1`](https://github.com/cloudius-systems/osv/commit/47ec18e1) by Christoph Hellwig
+- February 2019: OSv `vfs` (at commit [`eb28d3bd`](https://github.com/cloudius-systems/osv/commit/eb28d3bd)) is added to Unikraft's `vfscore` as commit [`54e65bbb`](https://github.com/unikraft/unikraft/commit/54e65bbb) by Yuri Volchcov
+- currently: 245 new commits with 9009 insertions and 5204 deletions have happend to Unikraft's `vfscore` implementation since the OSv import
```

Comment:
Maybe also add the date (not just currently).

Reply:
I would go also go for June 2023